roofit/histfactory/src/ParamHistFunc.cxx

```diff
@@ -521,83 +511,71 @@ RooArgList ParamHistFunc::createParamSet(const std::string& Prefix, Int_t numBin
 }
 
 
-////////////////////////////////////////////////////////////////////////////////
-/// return 0 for success
-/// return 1 for failure
-/// Check that the elements 
-/// are actually RooRealVar's
-/// If so, add them to the 
-/// list of vars
-Int_t ParamHistFunc::addVarSet( const RooArgList& vars ) {
-
-
-  int numVars = 0;
-
-  RooFIter varIter = vars.fwdIterator() ;
-  RooAbsArg* comp ;
-  while((comp = (RooAbsArg*) varIter.next())) {
-    if (!dynamic_cast<RooRealVar*>(comp)) {
-      coutE(InputArguments) << "ParamHistFunc::(" << GetName() << ") ERROR: component " 
-			    << comp->GetName() << " in variables list is not of type RooRealVar" 
-			    << std::endl;
-      RooErrorHandler::softAbort() ;
-      return 1;
-    }
-
-    _dataVars.add( *comp );
-    numVars++;
+/// Fill the mapping between RooDataHist bins and TH1 Bins.
+std::vector<int> const& ParamHistFunc::getParamSetBinMap() const {
 
+  if(!_binMapVector.empty()) {
+    return _binMapVector;
   }
 
-  Int_t numBinsX = 1;
-  Int_t numBinsY = 1;
-  Int_t numBinsZ = 1;
+  int numVars = _dataVars.size();
+
+  int numBinsX = 1;
+  int numBinsY = 1;
+  int numBinsZ = 1;
 
   if( numVars == 1 ) {
-    RooRealVar* varX = (RooRealVar*) _dataVars.at(0);
-    numBinsX = varX->numBins();
-    numBinsY = 1;
-    numBinsZ = 1;
+    numBinsX = static_cast<RooRealVar const&>(_dataVars[0]).numBins();
   } else  if( numVars == 2 ) {
-    RooRealVar* varX = (RooRealVar*) _dataVars.at(0);
-    RooRealVar* varY = (RooRealVar*) _dataVars.at(1);
-    numBinsX = varX->numBins();
-    numBinsY = varY->numBins();
-    numBinsZ = 1;
+    numBinsX = static_cast<RooRealVar const&>(_dataVars[0]).numBins();
+    numBinsY = static_cast<RooRealVar const&>(_dataVars[1]).numBins();
   } else  if( numVars == 3 ) {
-    RooRealVar* varX = (RooRealVar*) _dataVars.at(0);
-    RooRealVar* varY = (RooRealVar*) _dataVars.at(1);
-    RooRealVar* varZ = (RooRealVar*) _dataVars.at(2);
-    numBinsX = varX->numBins();
-    numBinsY = varY->numBins();
-    numBinsZ = varZ->numBins();
+    numBinsX = static_cast<RooRealVar const&>(_dataVars[0]).numBins();
+    numBinsY = static_cast<RooRealVar const&>(_dataVars[1]).numBins();
+    numBinsZ = static_cast<RooRealVar const&>(_dataVars[2]).numBins();
```

Comment:
Why not replace those three branches with code repetition by:
```c++
if (numVars >= 1) numBinsX = static_cast<RooRealVar const&>(_dataVars[0]).numBins();
if (numVars >= 2) numBinsY = static_cast<RooRealVar const&>(_dataVars[1]).numBins();
...
```
And I would assume that `numVars` > 0. You could check.

Reply:
Or if you want to go more RAII:
```c++
const int numBinsX = numVars >= 1 ? static_cast<RooRealVar const&>(_dataVars[0]).numBins() : 1;
...
```